File: Sean的leetcode/Leetcode/Nth Digit.cpp

class Solution {
public:
    int findNthDigit(int n) {
        int len = 1;
        int start = 1;
        long cnt = 9;
        while (n > cnt * len) {
            n -= cnt * len;
            len++;
            start *= 10;
            cnt *= 10;
        }
        start += (n - 1) / len;
        string num = to_string(start);
        return num[(n - 1) % len] - '0';
    }
};

File: Sean的leetcode/python/Nth Digit.py
###
# Back Tracking
# Time Complexity: O(log(n)) cnt is multiply by 10 each time
###
class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        l = 1
        start = 1
        cnt = 9
        while n > cnt * l:
            n -= cnt * l
            cnt *= 10
            start *= 10
            l += 1
        start += (n-1)/l
        start = str(start)
        return int(start[(n-1) % l])
        
        